### PR TITLE
Less latency

### PR DESCRIPTION
**Pull request by ryanpbrewster**
Trigger message ingestion after a message is sent
Trigger slack pushes after a message is ingested

Should lower latency from waiting for scheduled tasks.

@leogrim

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 1]
```diff
@@ -98,8 +100,19 @@ class LibraryToSlackChannelPusherImpl @Inject() (
...
-    libIds.foreach { libId => pushLibraryAtLatest(libId, nextPushAt) }
+    val now = clock.now
+    val libsById = libRepo.getActiveByIds(libIds)
+    val superFastOrgs = orgExperimentRepo.getOrganizationsByExperiment(OrganizationExperimentType.SLACK_COMMENT_MIRRORING).toSet
+    val (libsToPushImmediately, libsToPushSoon) = libIds.partition { libId =>
+      libsById.get(libId).exists(_.organizationId.exists(superFastOrgs.contains))
+    }
+    if (libsToPushImmediately.nonEmpty) {
+      libsToPushImmediately.foreach { libId => pushLibraryAtLatest(libId, now) }
+      session.onTransactionSuccess { pushingActor.ref ! IfYouCouldJustGoAhead }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: No need to use `onTransactionSuccess` when you control the transaction

**Comment by leogrim** [comment]
Re: Trigger message ingestion after a message is sent, is this needed if we're using Shoebox endpoints? Can we merely update the keep as we send the message in an idempotent way?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: I found it slightly cleaner than extracting the boolean from the db session. I'll try it the other way and see how it looks.

**Comment by ryanpbrewster** [comment]
It would not be needed, no. However, I imagine most messages are still being sent via eliza endpoints.

**Comment by leogrim** [comment]
I see, so we'd kill it as we move to Shoebox

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala]: I don't fully trust `onTransactionSuccess` (I wrote it), so I think we should only use it when there's no other option.

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/eliza/ShoeboxMessageIngestionActor.scala [hunk 2]
```diff
@@ -51,6 +57,9 @@ class ShoeboxMessageIngestionActor @Inject() (
       val messagesByKeep = messages.groupBy(_.keep)
       db.readWrite { implicit session =>
         val keeps = keepRepo.getByIds(messagesByKeep.keySet).values
+        session.onTransactionSuccess {
+          slackPusher.schedule(keeps.flatMap(_.connections.libraries).toSet)
+        }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/eliza/ShoeboxMessageIngestionActor.scala:62]: No need to use `onTransactionSuccess` here either

File: kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala [hunk 3]
```diff
@@ -97,9 +99,20 @@ class LibraryToSlackChannelPusherImpl @Inject() (
...
+      val (libsToPushImmediately, libsToPushSoon) = libIds.partition { libId =>
+        libsById.get(libId).exists(_.organizationId.exists(superFastOrgs.contains))
+      }
+      libsToPushImmediately.foreach { libId => pushLibraryAtLatest(libId, now) }
+      libsToPushSoon.foreach { libId => pushLibraryAtLatest(libId, now plus delayFromPushRequest) }
+
+      libsToPushImmediately.nonEmpty
+    }
+    if (weShouldPushImmediately) pushingActor.ref ! IfYouCouldJustGoAhead
   }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/LibraryToSlackChannelPusher.scala:116]: :+1:

**Pull request merged**